### PR TITLE
Expose Hosting Trial: Allow starting free trial from themes upload page

### PR DESCRIPTION
**Pull request by vykes-mac**
Related to https://github.com/Automattic/dotcom-forge/issues/5097

## Proposed Changes

This PR allow the user to start a free trial from the themes upload page.  There is an issue where we gets redirected to crowdsignal after the atomic transfer. what should happen is that we should be redirected to `/wp-admin/theme-install.php`

* Show free trial banner when user is eligible
* Shows the trial acknowledge modal and start the trial

## Testing Instructions

* go to `/themes/upload/simple_site`
* Start your free trial and verify the site transitions to atomic successfully 

image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by zaguiini** [comment]
>what should happen is that we should be redirected to /wp-admin/theme-install.php

Should we? What's wrong with keeping the user in that same page? They want to upload a theme, and they are in the right place in the sense that all they need to do is to upload the ZIP


After the transfer, it opened a ton of tabs:

image

**Comment by vykes-mac** [comment]
> Should we? What's wrong with keeping the user in that same page? They want to upload a theme, and they are in the right place in the sense that all they need to do is to upload the ZIP

Yeah it should redirect to `wp-admin`, it's a design decision that was made because the calypso upload doesn't allow updating a plugin so they are forcing the user to use the wp-admin one instead. See conversation  p1706055415216519-slack-C029FM1EH, If you go to /themes/upload on an AT site it will redirect you.

**Comment by vykes-mac** [comment]
`Should we? What's wrong with keeping the user in that same page? They want to upload a theme, and they are in the right place in the sense that all they need to do is to upload the ZIP`

Let's no redirect If we start the trial on this page. There's a bunch of redirect happening as it seems the atomic platform is not fully setup so `/theme-install.php`  returns 302 which intern redirect to another location and it continues until a valid `/wp-admin/` url is returned.

**Review by paulopmt1 — COMMENTED** [review]
The trial started successfully! After my site became Atomic, I still saw this screen and can't click on that area or drop zip files, it seems it is still disabled:

 image

After I refreshed it, it redirected me to the native theme page. Is that intentional?

image

**Comment by vykes-mac** [comment]
By design from previous comment it should redirect to wp-admin theme if I refresh. You should be able to upload a theme though without refreshing the page, I think what happened in this case the request site runs before the site is fully atomic on the backend so the isAtomic flag is still false and so the drop zone is disabled. I think I need to keep running requestSite until it returns isAtomic is true. I wait for 2 seconds before requesting site but sometimes it still seem to be too early

File: client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx [hunk 1]
```diff
@@ -24,29 +14,10 @@ export const WithOnclickTrialRequest = createHigherOrderComponent(
...
-			// until the `is_wpcom_atomic` = true is returned
-			if ( transferStatus === transferStates.COMPLETE && ! isSiteAtomic ) {
-				waitFor( 2 ).then( () => dispatch( requestSite( siteId ) ) );
-			}
-		}, [ transferStatus ] );
 
 		const fetchUpdatedData = () => {
 			//after transfer is complete we wait to fetch update site with `is_wpcom_atomic`
-			waitFor( 2 ).then( () => dispatch( requestSite( siteId ) ) );
+			waitFor( 3 ).then( () => dispatch( requestSite( siteId ) ) );
```
Comment (zaguiini) [inline client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx]: Would it make sense to poll this information instead of waiting an arbitrary number of seconds?

File: client/my-sites/plugins/plugin-upload/index.jsx [hunk 2]
```diff
@@ -203,9 +203,8 @@ const mapStateToProps = ( state ) => {
 	// before data has loaded.
 	const isEligible = isEligibleForAutomatedTransfer( state, siteId );
 	const isEligibleForHostingTrial =
-		isUserEligibleForFreeHostingTrial( state ) &&
-		site.plan.is_free &&
-		config.isEnabled( 'hosting-trial' );
+		isUserEligibleForFreeHostingTrial( state ) && site && site.plan?.is_free;
+	config.isEnabled( 'hosting-trial' );
```
Comment (zaguiini) [inline client/my-sites/plugins/plugin-upload/index.jsx]: What's happening here? The trial check went over to a new line :) Can we also replace the site check with `site?.plan?.is_free`?

File: client/my-sites/themes/theme-upload/index.jsx [hunk 3]
```diff
@@ -205,8 +249,16 @@ class Upload extends Component {
 		tryandcustomize.action( this.props.themeId );
 	};
 
+	onUpsellNudgeClick = () => {
+		const isEligibleForHostingTrial = this.props.isEligibleForHostingTrial;
```
Comment (zaguiini) [inline client/my-sites/themes/theme-upload/index.jsx]: That's a funny assignment. We're not destructuring nor renaming the var 🤔 What was the intent here?

File: client/my-sites/themes/theme-upload/index.jsx [hunk 4]
```diff
@@ -223,6 +275,17 @@ class Upload extends Component {
...
 
+		if ( isEligibleForHostingTrial ) {
+			/* translators: %(planName)s the short-hand version of the Business plan name */
+			title = translate(
+				'Start your free %(planName)s plan trial to access the theme install features',
+				{
+					args: { planName: getPlan( PLAN_BUSINESS )?.getTitle() ?? '' },
+				}
+			);
+			upgradeUrl = '#';
```
Comment (zaguiini) [inline client/my-sites/themes/theme-upload/index.jsx:285]: Do we absolutely need to pass the `#`?

File: client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx (hunk 1 above)
Comment (vykes-mac) [reply client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx]: Yup that's the intention!

File: client/my-sites/themes/theme-upload/index.jsx (hunk 4 above)
Comment (vykes-mac) [reply client/my-sites/themes/theme-upload/index.jsx:285]: If we don't the upsell internals gonna add plans url

Comment (vykes-mac) [reply client/my-sites/themes/theme-upload/index.jsx:285]: unless we change the internals

File: client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx (hunk 1 above)
Comment (zaguiini) [reply client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx]: I see. The `waitFor` function dispatches a timeout handler, so the dispatch is only dispatched once after three seconds. Are we pulling somewhere else?

File: client/my-sites/themes/theme-upload/index.jsx (hunk 3 above)
Comment (vykes-mac) [reply client/my-sites/themes/theme-upload/index.jsx]: intent is stupid mistake 😓

File: client/my-sites/themes/theme-upload/index.jsx (hunk 4 above)
Comment (zaguiini) [reply client/my-sites/themes/theme-upload/index.jsx:285]: Seems like it's trickier to change the internals, but I only glanced over the code.

File: client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx (hunk 1 above)
Comment (vykes-mac) [reply client/my-sites/plans/trials/trial-acknowledge/with-onclick-trial-request.tsx]: We poll in the `<HostingActivateStatus>` component. but we don't do any dispatch there, keeping the relevent functions to be dispatched out of that component, the component only returns the status.

File: client/my-sites/themes/theme-upload/index.jsx (hunk 4 above)
Comment (vykes-mac) [reply client/my-sites/themes/theme-upload/index.jsx:285]: Yeah I didn't want to go down that rabbit hole

Comment (zaguiini) [reply client/my-sites/themes/theme-upload/index.jsx:285]: Sounds good 👍

**Review by zaguiini — APPROVED** [review]
Works well!

File: client/my-sites/plugins/plugin-upload/index.jsx [hunk 5]
```diff
@@ -204,7 +203,7 @@ const mapStateToProps = ( state ) => {
 	const isEligible = isEligibleForAutomatedTransfer( state, siteId );
 	const isEligibleForHostingTrial =
 		isUserEligibleForFreeHostingTrial( state ) && site && site.plan?.is_free;
-	config.isEnabled( 'hosting-trial' );
```
Comment (vykes-mac) [inline client/my-sites/plugins/plugin-upload/index.jsx]: @paulopmt1 we need to revert this file to trunk. We will make these plugins related changes in another PR

Comment (vykes-mac) [reply client/my-sites/plugins/plugin-upload/index.jsx]: Here's the PR https://github.com/Automattic/wp-calypso/pull/86968

**Review by paulopmt1 — APPROVED** [review]
It's working as expected, nice work @vykes-mac! I was able to start my trial, submit a theme file and activate it. All at the same flow!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/11099086

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @vykes-mac for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/11099086).
